From af5d32343ffef1d17b4b83199934ddf426c8c507 Mon Sep 17 00:00:00 2001
From: whitebelyash <whbexiumwork@gmail.com>
Date: Fri, 16 Jan 2026 22:41:23 +0400
Subject: [PATCH 1/2] HACK: u_gralloc: always use ubwc detection path

u_gralloc checks if the current buffer was allocated by the Qualcomm vendor gralloc by comparing 'gmsm' magic with the buffer data. Since newer gralloc versions this way isn't working anymore, causing the condition to never success - essentially skipping setting proper DRM format modifier. This caused huge distorted picture on the screen because the allocated buffer never got a proper modifier as it should have

Signed-off-by: whitebelyash <whbexiumwork@gmail.com>
---
 src/util/u_gralloc/u_gralloc_fallback.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/src/util/u_gralloc/u_gralloc_fallback.c b/src/util/u_gralloc/u_gralloc_fallback.c
index 44fb32d8cfd..bb6459c2e29 100644
--- a/src/util/u_gralloc/u_gralloc_fallback.c
+++ b/src/util/u_gralloc/u_gralloc_fallback.c
@@ -148,12 +148,16 @@ fallback_gralloc_get_buffer_info(struct u_gralloc *gralloc,
    out->strides[0] = stride;

 #ifdef HAS_FREEDRENO
-   uint32_t gmsm = ('g' << 24) | ('m' << 16) | ('s' << 8) | 'm';
+  /* uint32_t gmsm = ('g' << 24) | ('m' << 16) | ('s' << 8) | 'm';
    if (hnd->handle->numInts >= 2 && hnd->handle->data[hnd->handle->numFds] == gmsm) {
-      /* This UBWC flag was introduced in a5xx. */
-      bool ubwc = hnd->handle->data[hnd->handle->numFds + 1] & 0x08000000;
-      out->modifier = ubwc ? DRM_FORMAT_MOD_QCOM_COMPRESSED : DRM_FORMAT_MOD_LINEAR;
+
    }
+   */
+
+   /* TODO: Actually find a way to detect a Qualcomm vendor allocated buffer */
+   /* This UBWC flag was introduced in a5xx. */
+   bool ubwc = hnd->handle->data[hnd->handle->numFds + 1] & 0x08000000;
+   out->modifier = ubwc ? DRM_FORMAT_MOD_QCOM_COMPRESSED : DRM_FORMAT_MOD_LINEAR;
 #endif

    return 0;
--
2.53.0

From 50e50fac699b4921d3574d8fb96e8371a80cd684 Mon Sep 17 00:00:00 2001
From: whitebelyash <whbexiumwork@gmail.com>
Date: Mon, 2 Mar 2026 13:34:05 +0400
Subject: [PATCH 2/2] freedreno/shim: enable drm-shim support for Adreno
 810/825/829

Cache sizes for 810&825 should be valid. 829 is unknown, but works so
far(?)

Signed-off-by: whitebelyash <whbexiumwork@gmail.com>
---
 src/freedreno/drm-shim/freedreno_noop.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/src/freedreno/drm-shim/freedreno_noop.c b/src/freedreno/drm-shim/freedreno_noop.c
index 4c3e817fce7..fff7f4f3251 100644
--- a/src/freedreno/drm-shim/freedreno_noop.c
+++ b/src/freedreno/drm-shim/freedreno_noop.c
@@ -281,6 +281,21 @@ static const struct msm_device_info device_infos[] = {
       .chip_id = 0x44050000,
       .gmem_size = 12 * 1024 * 1024,
    },
+   {
+      .gpu_id = 825,
+      .chip_id = 0x44030000,
+      .gmem_size = 2 * 1024 * 1024,
+   },
+   {
+      .gpu_id = 829,
+      .chip_id = 0x44030A20,
+      .gmem_size = 2 * 1024 * 1024,
+   },
+   {
+      .gpu_id = 810,
+      .chip_id = 0x44010000,
+      .gmem_size = 576 * 1024,
+   }
 };

 static void
--
2.53.0
